libs/sdk-core/src/error.rs

```diff
@@ -1,60 +1,311 @@
+pub enum ServiceErrorCode {
+    BackupFailed,
+    ClientError,
+    DerivationFailed,
+    Generic,
+    InitFailed,
+    NotReady,
+    PersistenceFailed,
+    RequestFailed,
+    ShutdownFailed,
+    SignatureError,
+    StreamAlreadyCreated,
+    TransportError,
+}
+
+impl fmt::Display for ServiceErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// The LSP error codes
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
+pub enum LspErrorCode {
+    CalculateOpenChannelFeesFailed,
+    ChannelOpeningNotAvailable,
+    ConnectionFailed,
+    LspNotFound,
+}
+
+impl fmt::Display for LspErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// The payment error codes
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
+pub enum PaymentErrorCode {
+    AlreadyPaid,
+    ChannelNotFound,
+    InvalidAmount,
+    OpeningFeeParamsNotFound,
+    PaymentFailed,
+    PaymentNotFound,
+    PaymentTimeout,
+}
+
+impl fmt::Display for PaymentErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// The onchain error codes
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
+pub enum OnchainErrorCode {
+    InvalidAddress,
+    InvalidAmount,
+    InvalidPaymentHash,
+    InvalidSwap,
+    SwapAddressNotFound,
+    SwapInProgress,
+    SwapInsufficientFunds,
+    SwapMissingInput,
+    SwapProviderError,
+    UnexpectedAddress,
+    UnexpectedScript,
+}
+
+impl fmt::Display for OnchainErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+/// The value error codes
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
+pub enum ValueErrorCode {
+    ConversionFailed,
+    DecodingFailed,
+    InvalidMnemonic,
+    InvalidSignature,
+    InvalidTimestamp,
+    InvalidUrl,
+    MissingParameters,
+    UnexpectedLength,
+    UnexpectedType,
+    UnknownValue,
+}
+
+impl fmt::Display for ValueErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
 /// Type of error returned by the SDK
 #[derive(Debug, Error)]
 pub enum SdkError {
-    /// Generic error, that doesn't fit any of the other types
-    #[error("Breez SDK error: {err}")]
-    Generic { err: String },
+    #[error("Service error [{code}]: {err}")]
+    ServiceError { code: ServiceErrorCode, err: String },
+
+    #[error("LSP error [{code}]: {err}")]
+    LspError { code: LspErrorCode, err: String },
+
+    #[error("Payment error [{code}]: {err}")]
+    PaymentError { code: PaymentErrorCode, err: String },
+
+    #[error("Onchain error [{code}]: {err}")]
+    OnchainError { code: OnchainErrorCode, err: String },
+
+    #[error("Value error [{code}]: {err}")]
+    ValueError { code: ValueErrorCode, err: String },
+}
+
+impl SdkError {
```

Comment:
NIT: Can the methods that take `err: String` as argument be changed to take instead `err: &str` and do the conversion to `String` internally?

This would make calling them easier, because we would avoid calling `to_string()` every time we give in an error message.

```rust
ensure_sdk!(condition, SdkError::generic("Danger error"));
```

If we happen to have a `String`, we can then give it as a reference

```rust
let error_msg: String = "Some message".
ensure_sdk!(condition, SdkError::generic(&error_msg));
```

Reply:
Done [4edbbbe](https://github.com/breez/breez-sdk/pull/469/commits/4edbbbee20fdca35aebe326f3048148e0f9c7b48). In most cases Strings are now referenced like below. Let me know if there is a more rusty way to do it.
```
SdkError::lsp_error(LspErrorCode::ConnectionFailed, &e.to_string())
SdkError::lsp_not_found(&format!("Unknown LSP: {lsp_id}"))
```

Reply:
There might be fancier ways, but this is very good!